### DISCUSSION TITLE
Extend .NET MAUI application with iOS Extensions

### DISCUSSION THREAD

**Discussion by giscus** [VladislavAntonyuk/vladislavantonyuk.github.io#72] [Announcements]
# Extend .NET MAUI application with iOS Extensions

Extend the .NET MAUI application with iOS App Extension in 5 steps.

https://vladislavantonyuk.github.io/articles/Extend-.NET-MAUI-application-with-iOS-Extensions/

**Comment by vassilych** [comment]
Hi, very nice trick to add your app to the list of apps to share files from other apps.
Do you have a full example? Including your app receiving and processing the file sent by another app?
And all necessary additional configs for that in info.plist

Thanks, Vassili

**Reply by VladislavAntonyuk** [reply]
Hello, see https://github.com/VladislavAntonyuk/ContinueOnPC
It receives url and send it to the firebase.

**Comment by vikher** [comment]
Is there any guidance on generating an IPA build locally using the dotnet publish command, given the challenge of handling two projects when Maui is expected to be a single project? Specifically, considering the need for a separate provisioning profile for the iOS extension, how can the command accommodate this limitation since it appears to accept only one provisioning profile using parameters such as:
dotnet publish -f $(iOSVersion) -c $(BuildConfiguration) /p:ArchiveOnBuild=true /p:RuntimeIdentifier=ios-arm64 /p:EnableAssemblyILStripping=false /p:CodesignKey=$(CodesignKey) /p:CodesignProvision=$(CodesignProvision)
The issue lies in managing the requirements of two distinct projects within the context of Maui, which is expected to be a unified project. Specifically, the iOS extension necessitates its own provisioning profile. Given that the dotnet publish command seems to allow only one provisioning profile, how can this limitation be addressed to successfully build the project?

**Reply by VladislavAntonyuk** [reply]
I don't have Apple developer account to give you the right answer, but the arguments should be the same as you have in xamarin msbuild.

**Reply by VladislavAntonyuk** [reply]
Does automatic provisioning work?

Also you can put these settings directly in csproj file and dotnet publish should resolve them.

**Comment by borigas** [comment]
Thanks for writing this and the MAUI templates. You're filling a big hole in the Xamarin Forms -> MAUI migration story.

I followed your post here and your notification service extension template to build https://github.com/borigas/NotificationServiceExtensionRepro, but I get an error trying to deploy. "Please select a valid device before running the application". I posted details in an issue at https://github.com/dotnet/maui/issues/22586. 

Does this still work for you? Did the tooling regress at some point? Did I miss some detail in how I set this up?

Thanks again for your help!

**Reply by VladislavAntonyuk** [reply]
I will check, but as you can see from the video in the article it worked at the moment of writing.

**Reply by borigas** [reply]
Agreed. It seems like multiple people have had good success using your templates and write up here. I didn't intend to imply that it didn't work when the article was written.

But I spent weeks trying to get it to working, trying multiple versions of VS, VSCode, XCode, different devices, simulators, various csproj settings, and can't get it to deploy. I'm hoping I misconfigured something in the above repro repo, but I'm afraid some VS/Maui tooling regressed. I'm having trouble getting traction with those teams, so trying to further validate that it's not some silly mistake on my end.

**Reply by borigas** [reply]
I don't mean to nag, but have you had a chance to test this with current versions?

**Comment by tomzeni** [comment]
Hi, could you please add the new iOS 18 ContactsProvider Extension Template (https://developer.apple.com/documentation/ContactProvider)  
to you collection of Maui iOS Extension Templates at:: https://www.nuget.org/packages/VladislavAntonyuk.DotNetTemplates.
Thanks a lot, Tom

**Comment by tkefauver** [comment]
Any idea how this could work with Avalonia?

**Reply by VladislavAntonyuk** [reply]
the same. There are no .NET MAUI dependencies here

**Comment by Cedric1235** [comment]
Hello, I'm working with .NET 8 and trying to create an iOS Notification Extension Service using .NET MAUI. However, I'm encountering some issues that I can't seem to resolve. Specifically, on my Mac, it's showing an "invalid configuration mapping" error next to the Notification Extension Service. Can you help me with that?

**Comment by alpayclsr1** [comment]
Thank you for making great contents for .net maui . According to the documentation you prepared for the iOS share extension in .NET MAUI, I implemented the share extension. However, when I run the application, it deploys the share extension project to iOS, even though the startup project is set to the MAUI application. To test, I downloaded the Joke app that you prepared, but I encountered the same issue— it deployed the widget app to the phone. Can you help me identify where I am going wrong?


Thanks

**Reply by VladislavAntonyuk** [reply]
That is correct. You can’t deploy only extension, since extension is a part of the app.
To deploy only app without extension just remove the reference to the extension.

**Reply by alpayclsr1** [reply]
Thanks for response I am not running the extension project directly. I am running the .net maui(Joke app), but it still deploys the extension project

**Reply by VladislavAntonyuk** [reply]
As I mentioned that is correct, because extension is part of the app. To deploy only app, just remove reference to the extension

**Reply by Pelinalpp** [reply]
Hello VladislavAntonyuk, I have the same problem as Alpay. Actually, I know that when I remove the reference of the Share Extension project, the main project will be loaded. However, what I want is to run the maui project without losing the Share Extension feature. If we remove the reference of the share extension project from the maui project, won't we lose the share extension features? I took a screen recording of what I meant. Since I couldn't upload a video here, I uploaded the video to the link below. I think you'll understand what I mean when you watch it. (The project I am running in the screen record is your share extension sample. I run AppContainer project as startup project, but I see that the AppContainerExtension project installed in de device. Also, the project does not open.)

https://drive.google.com/file/d/1oqu0-Sze15aTIgY0-OjFhLCBcN_KjRkw/view?usp=sharing

**Reply by VladislavAntonyuk** [reply]
Could you please attach your sample app?
I didn't try it in VSCode, maybe the issue is in IDE. The dotnet build should work and deploy the correct app. The extension itself doesn't have an icon, so something is wrong with the deployed project.

**Reply by Pelinalpp** [reply]
Sample project is your sample project in your GitHub. I downloaded and test it. But i guess it is bug of .NET MAUI Debugger. I picked .NET Meteor Debugger as Debugger. And I deployed it though .NET Meteor Debugger in iPhone simulator, it run successfully. Thanks.

**Comment by LucaDuttoTD** [comment]
Great work, thanks! I'm using your MAUI templates at the moment.

I have a question tho:
In my project I add a project reference

```
	<ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'">
	  <ProjectReference Include="..\xxx\xxx.csproj" >
			<IsAppExtension>true</IsAppExtension>
			<IsWatchApp>false</IsWatchApp>
		</ProjectReference>
	</ItemGroup>
```
but the reference in the solution explorer is marked with a yellow warning icon, and I see a dll is not created in the bin folder for ios...
is that the correct behaviour?

Thanks!

**Reply by VladislavAntonyuk** [reply]
Hello, @LucaDuttoTD. No, seems like there is an issue with target framework compatibility. Try build extension project without main app. Also check if your extension target framework matches the target framework of your main app.

**Reply by LucaDuttoTD** [reply]
Thank you for your quick reply!

If I compile just the extension project everything is ok.
Target framework is net8.0-ios for the extension, net8.0-android;net8.0-ios for main app, it looks good to me.
I conditionally compile it only for ios, as you can see in my previous message

adding some infos: if i remove the following lines from main app csproj

```
<IsAppExtension>true</IsAppExtension>
<IsWatchApp>false</IsWatchApp>
```
the warning disappears...

**Reply by VladislavAntonyuk** [reply]
Everything looks correct.
I can take a look if you create a small reproduction sample. If you try MauiSamples app, do you see the same issue? I use ShareExtension there.
What extension type do you use?

**Reply by LucaDuttoTD** [reply]
Hi, and thank you for your support.

I created a minimal sample project reproducing the issue: https://github.com/technicaldesignit/TestApp
I'm using VS2022 latest preview version and .NET 8, if that helps

I'm trying the Notification Service Extension

Opening up you MauiSamples project I see the same identical issue in iOSExtensions/JokeApp: the JokeAppWidget reference is "yellow"

**Reply by VladislavAntonyuk** [reply]
I also see the same warning. I will take a look. Something may have changed in the recent .NET MAUI versions.

**Reply by VladislavAntonyuk** [reply]
The issue is gone with .NET 9 release:
![image](https://github.com/user-attachments/assets/66c74b18-b53b-41bf-8b55-6db4caa356f4)

**Reply by winstonpang** [reply]
I've just tried this in VS2022 17.12.2, with .NET 9 SDK installed, but the IsAppExtension is still not working anymore.

**Comment by ai-neko** [comment]
I've recently tried building and deploying this from a Mac (using Rider) to an iPhone.
I had no issues deploying the app and it also shows up as a share target. However, when I try to share something (a url from safari in my case) the expected UI  doesn't show up and the extension fails.
I looked through the console logs and while there's a lot of things related to the extension, I couldn't find the root cause.
Any tips or ideas? Does the extension still work for you?

**Reply by VladislavAntonyuk** [reply]
I checked with .NET 8. I will take a look with .NET 9

**Reply by VladislavAntonyuk** [reply]
I've just verified. Works as expected on iOS Simulator.
<img width="1800" alt="Screenshot 2025-01-19 at 15 59 26" src="https://github.com/user-attachments/assets/c2269dfa-18b7-4273-9b3b-77d3a148d5e0" />
<img width="368" alt="Screenshot 2025-01-19 at 16 00 26" src="https://github.com/user-attachments/assets/0c0c4900-3954-4f95-8979-4348dd4f11c6" />

**Reply by ai-neko** [reply]
Thank You for looking into this.

I can confirm that the app works in the emulator for (iOS 18.2).
However, when I try the same thing on my actual phone, (iPhone 15, iOS 18.2.1) the extension fails. 
I've re-tested it and as before, the app deploys correctly, the share target shows up but when it's selected, the screen just blinks and the screen to actually post the link does not show up.

**Reply by VladislavAntonyuk** [reply]
I can only check on my iPhone 8 (iOS 16.7.10), hopefully it will not depend on iOS version.

**Reply by EmilienDup** [reply]
Hi there,

I am also facing the same problem.
Everything works fine when deployed in the simulator, but when deployed on a real device, the extension UI can never be displayed.

Tested on an iPad mini 4  (iOS 15.8.2), and an iPhone 14 (iOS 18.3).

App built using Rider, XCode 16.4, and latest dotnet workloads (https://github.com/xamarin/xamarin-macios/releases/tag/dotnet-9.0.1xx-xcode16.2-9170).

As it seems not to be any official MAUI documentation for iOS extensions, it is hard to tell if the problem lies with the sample app, the build toolchain, or with maui/xamarin-macios itself.

If I can assist in any way to investigate this problem, please let me know.

Thanks a lot.

**Reply by VladislavAntonyuk** [reply]
Adding this code in Extension project fixes the issue:
```
	<PropertyGroup>
		<Registrar>static</Registrar>
	</PropertyGroup>
```

I will update templates

**Reply by EmilienDup** [reply]
I've tested it on my side, and I can confirm that adding the `Registrar` property fixed the problem.

Thanks.

**Comment by SergTomcat** [comment]
I don't know if I have come to the right place, but maybe someone could help me.

I tried to create Notification Service folowing this tutorial, but I'm getting "*Xamarin.iOS does not support deploying the previous version. Please ensure your solution builds before deploying it.*"  error when deploying on local iPhone and "*The source '../../IOSNotificationServiceExtension/5cdf4224e9b3649828749b08d1926d9385e09b70cfcb8e952173e6928ddf70bb/bin/Debug/net9.0-ios/ios-arm64/device-builds/iphone10.6-16.7.8/IOSNotificationServiceExtension.appex' does not exist.*" when I try to build solution on remote mac.

.Net 9, tried every suggestion, even this ones: https://developercommunity.visualstudio.com/t/Cannot-build-MAUI-or-iOS-NET-project-wh/10582380

**Reply by VladislavAntonyuk** [reply]
ditto exit code means your Xcode version is incompatible with .NET for iOS. Update to .NET 9 and Xcode 16.2, that should help.

**Reply by SergTomcat** [reply]
Unfortunately it does not.
I managed to publish app to TestFlight (create all certs and profiles, build release, copy archive on mac and then publish with XCode, because VS cannot properly link signing identities to extension).
But when I try to deploy debug build (both on local iPhone and on remote iPhone, connected to mac), I get "Xamarin.iOS does not support..." error in first case and "IOSNotificationServiceExtension.appex' does not exist." in the second one.

Did anyone managed to run debug build with extension?

**Comment by NguyenTuanAnh0708** [comment]
Can you setup with net 9 ??

**Reply by VladislavAntonyuk** [reply]
I updated VladislavAntonyuk.NET-Templates NuGet package that supports net9

**Reply by NguyenTuanAnh0708** [reply]
Sorry but can you tell me where to find that NuGet

**Reply by VladislavAntonyuk** [reply]
Source with link to NuGet https://github.com/VladislavAntonyuk/.NET-Templates

**Reply by NguyenTuanAnh0708** [reply]
I have two questions:
– How can I retrieve data from the Share Extension in the main app?
– How can I customize the layout of the Share Extension?

**Reply by VladislavAntonyuk** [reply]
Treat extension as a separate app. You can pass some data between them. If you want to pass data to your app you can use app links. It will start your main application. You can see my repo ContinueOnPC. It uses share extension to pass data

**Reply by NguyenTuanAnh0708** [reply]
so can i customize this layout Share Extension ?

**Reply by VladislavAntonyuk** [reply]
Yes, https://medium.com/@henribredtprivat/create-an-ios-share-extension-with-custom-ui-in-swift-and-swiftui-2023-6cf069dc1209

**Reply by NguyenTuanAnh0708** [reply]
This is my DidSelectPost function, but it doesn't open the deep link. I copied your code, but it doesn't seem to work for me. I'm not sure if I need to set up anything else (I'm sure the deep link is configured correctly).
      public override void DidSelectPost()
        {
            if (ExtensionContext is null)
            {
                return;
            }

            foreach (
                var attachment in ExtensionContext
                    .InputItems.Where(extensionItem => extensionItem.Attachments is not null)
                    .SelectMany(extensionItem => extensionItem.Attachments!)
                    .Where(attachment => attachment.HasItemConformingTo(UTTypes.Url.Identifier))
            )
            {
                attachment.LoadItem(
                    UTTypes.Url.Identifier,
                    null,
                    async (data, error) =>
                    {
                        var nsUrl = data as NSUrl;
                        var url = nsUrl?.AbsoluteString;
                        if (url is not null)
                        {
                            var escapedUrl = Uri.EscapeDataString(url);
                            var deepLink = $"yourapp://share?url={escapedUrl}";
                            var sharedUrl = new NSUrl(deepLink);

                            var success = await UIApplication.SharedApplication.OpenUrlAsync(
                                sharedUrl,
                                new UIApplicationOpenUrlOptions()
                            );

                            Console.WriteLine($"OpenUrlAsync success: {success}");
                        }
                    }
                );
            }

            // Inform the host that we're done, so it un-blocks its UI. Note: Alternatively you could call super's -didSelectPost, which will similarly complete the extension context.
            ExtensionContext.CompleteRequest(Array.Empty<NSExtensionItem>(), null);
        }

**Reply by NguyenTuanAnh0708** [reply]
this is repo : https://github.com/NguyenTuanAnh0708/share-extentions

**Comment by LionBlackk** [comment]
In MAUI, action extension which working on simulator, but it's not working on device.

**Reply by LionBlackk** [reply]
now, i use .NET 8

**Reply by LionBlackk** [reply]
i did set up
app .csproj
```
    <ItemGroup>
      <ProjectReference Include="..\MyHiveActionExtension\MyHiveActionExtension.csproj">
        <IsAppExtension>true</IsAppExtension>
        <IsWatchApp>false</IsWatchApp>
      </ProjectReference>
    </ItemGroup>
```
extension pj
```
<ApplicationId>com.dtp.mycontacts.ActionExtension</ApplicationId>
```
Info.plist extension
```
<string>com.dtp.mycontacts.ActionExtension</string>
```

**Reply by LionBlackk** [reply]
currently, on my device button action is not showing, but on my simulator, it's showing

**Reply by VladislavAntonyuk** [reply]
Most likely, the extension is not deployed to the device.

**Reply by LionBlackk** [reply]
I simply want to save the file to my current app, is there a way to solve this?

**Comment by andersonvieiragomeslopes** [comment]
Does anyone have any idea how to implement iOS Dynamic Island using an ActivityKit widget in a .NET MAUI application or sample?
Is the approach similar to how you would implement a standard widget, or is it significantly different?
Any guidance, code samples, or documentation would be greatly appreciated.
Also, is it currently possible to integrate Live Activities/Dynamic Island features in MAUI, or would it require creating a native extension?

**Comment by Szin** [comment]
Hi! I am so glad to find a source for using iOS Extensions in a MAUI App. 

Struggling for weeks so far to get an Autofill Credential Provider extension running under MAUI (already running in an Xamarin App) At least building and deploing would be a big step forward.

So I tried to get the ShareExtension from your MauiSamples repository (the iOSExtensions/ShareExtension projects) running, but build failed with:

The source '../../AppContainerExtension/b9176f09788f1a6d33df8a0b5a8c8c3761540c4b762c29e976b0d4451aca65ce/bin/Debug/net10.0-ios/iossimulator-arm64/AppContainerExtension.appex' does not exist.

I am using latest XCode 26, newest VS2026, .net10 RC2 on both machines

Can you please help? I would be very grateful!

**Reply by VladislavAntonyuk** [reply]
Hello, unfortunately integration with VS on Windows and MacOS doesn't work.
You need to run it directly on MacOS.

**Reply by Szin** [reply]
Thank you for your advice! I got an entry in the Password Autofill OS Settings now. Next trouble is to get up an MAUI UI when the extension is requesting actions. I tried to embed MAUI but run into issues
https://learn.microsoft.com/en-us/dotnet/maui/platform-integration/native-embedding?view=net-maui-10.0&pivots=devices-ios
When building the MAUI App instance i got
"Platform application instance is required and was not able to be detected."
I know that is not part of the article here but maybe you run in an similar situation where no "Containing app" is available.
Cheers and thankyou for your good work!

**Reply by VladislavAntonyuk** [reply]
If you trying to use MAUI in extension, I am not sure if it is possible. Probably the issue you mentioned said it expects the full app instance, but it is not detected (maybe extension process is not a full app instance)
